データベース設計

hi-roto opened the pull request:
# what
   データベース設計

# why
   アプリ作成のを効率的に行うため

# ER図
　https://gyazo.com/e2cef90b6943b203145f2382cfca2d2c

hi-roto committed:
実装準備

gem "robocop"導入
.ds_storeの設定
turbolinkの解除

hi-roto committed:
README作成

chisesakai commented on README.md:
```diff
@@ -1,24 +1,87 @@
+- has_many :entries
 
-* Database initialization
+## entries テーブル
```
entries テーブルとpurchases テーブルを統合して、
「購入情報管理をするテーブル」を作成しましょう！
下記をご参考くださいませ。

保存するもの
- user(購入者)のid
- item(購入する商品)のid

アソシエーション
- user(idを保存してるため)
- item(idを保存してるため)
- 住所テーブルとも紐づく

hi-roto replied:
レビューありがとうございます。
指摘事項を変更しました。下記URLが変更分のコミットです。
https://github.com/hi-roto/furima-32883/pull/1/commits/aadbeaa6d713c3bd68172acdb0c3ea75b976f613
アソシエーションについてですが、purchase_recordsテーブルとitemsテーブルの関係は、１対１という解釈でよろしいでしょうか？
１人の購入者は１つの商品を購入している、１つの商品は１人の購入者が存在すると理解しました。
修正したER図です。
https://gyazo.com/feda6aa78062b6718e33b61ecdbb8506

chisesakai commented on README.md:
```diff
@@ -1,24 +1,87 @@
+- has_many :entries
+- has_one :shipping
+
+## shipping テーブル
```
shippingテーブルとshipping_buildingテーブルを統合し、
「住所を保存するテーブル」を作成しましょう！
「購入情報管理をするテーブル」と紐づきます。

hi-roto replied:
レビューありがとうございます。
こちらの指摘事項について、修正させていただきました。下記URLが変更分のコミットです。
https://github.com/hi-roto/furima-32883/pull/1/commits/4676ad924b74ff7d01d60911da340c4efc6f65b7

確認させていただきたいのですが、私は購入画面において、building_name項目は必須項目ではないので、not null制約にならない。だから、このままshipping_adressesテーブルにて、カラムを作成した場合、空白欄が存在することになると考えました。
そこで、必須項目ではないカラムについて別テーブルを作成して、そこにデータを保存し、空白欄がないように設計したのですが、この解釈は間違っているのでしょうか？
また、カリキュラムのヒント
https://master.tech-camp.in/curriculums/5439の「手順２」
には、カラムを設定する際は「必須項目」が保存すべきカラムであると記載がありましたが、仮にデータベースに値を保存しない場合、実際に商品を購入して、購入者が入力した住所に配送する場合はどのように必須項目ではない（データベースに保存されない、ここで言う「建物名」）配送先を特定することになるのでしょうか？
修正前のER図です。https://gyazo.com/d7c586e2d6df8b9de30c3981f0bc3b43

長文で申し訳ありませんが、よろしくお願いします。

chisesakai commented on README.md:
```diff
@@ -1,24 +1,87 @@
+## entries テーブル
 
-* How to run the test suite
+| Column      | Type       | Options                        |
```
外部キーを保存するカラム名は、`_id`を削除しましょう！
references型の命名規則に則るためです。

hi-roto replied:
修正完了しました。

chisesakai commented on README.md:
```diff
@@ -1,24 +1,87 @@
+| region             | string     | null: false                    |
+| preperration_day_id| integer    | null: false                    |
+| price              | integer    | null: false                    |
+| saler_id           | references | null: false, foreign_key: true |
```
```
| saler_id           | references | null: false, foreign_key: true |
```
こちら、実際に紐づくテーブル名である`user`に変更しておきましょう！

hi-roto replied:
修正完了しました。
カラム名の変更とreferenceses型の命名規則を変更しています。
https://github.com/hi-roto/furima-32883/pull/1/commits/4801d50eb171c46ac85d93484879e8c7ff437ba9

chisesakai reviewed (commented):
大変お待たせ致しました！レビュー依頼ありがとうございます。
キャプチャ・コードともに拝見させていただきました。
#### 4件コメントいたしましたので、ご確認をお願いいたします。
いよいよ最終課題ですね、今後とも一緒に頑張っていきましょう！

hi-roto committed:
entriesテーブルとpurchasesテーブルの統合

１entriesテーブルとpurchasesテーブルの統合
２アソシエーションの変更
３purchasesテーブルをpurchase_recordsに名称変更
４shippingテーブルをshipping_adressesに名称変更

hi-roto committed:
shipping_buildingテーブルの統合

1 shipping_buildingテーブルを削除し、purchase_adressesテーブルに統合
2 purchase_adressesテーブルにbuilding_nameカラムを追加

hi-roto committed:
カラム名の変更

sailer_idカラムをuserに変更

hi-roto commented:
レビューありがとうございました。
これからよろしくお願いします。頑張ります。
指摘事項について上記のとおり修正しました。
よろしくお願いします。
完成ER図です。
https://gyazo.com/79f0a892053cf162f4113a93ec5df23d

Hide-Morishita commented on README.md:
```diff
@@ -1,24 +1,63 @@
+| ----------------| ------ | -------------------------|
+| nickname        | string | null: false              |
+| email           | string | null: false, unique: true|
+| password        | string | null: false              |
```
```ruby
| password        | string | null: false              |
```
deviseのgemを使用しての実装になるため、passwordカラムのカラム名はdeviseが予め用意しているencrypted_passwordにしましょう。

hi-roto replied:
修正しました。
アプリの仕様を確認したら、ニックネームも一意制限がかかっていたので、オプションに追加しました。
25659b2

Hide-Morishita commented on README.md:
```diff
@@ -1,24 +1,63 @@
+| category_id        | integer    | null: false                    |
+| item_status_id     | integer    | null: false                    |
+| postage_payer_id   | integer    | null: false                    |
+| region             | string     | null: false                    |
```
```ruby
| region             | string     | null: false                    |
```
こちらのカラムは「発送元の地域(都道府県)」のカラムでしょうか？
「発送元の地域(都道府県)」のカラムであれば、shipping_adressesテーブルのprefecture_idと同じactive_hashを使用できますので、統一しましょう。
```ruby
| prefecture_id    | integer    | null: false                    |
```

hi-roto replied:
regionは発送元の地域のことです。
修正しました。
6621f61

Hide-Morishita commented on README.md:
```diff
@@ -1,24 +1,63 @@
+
+| Column           | Type       | Options                        |
+| -----------------| ---------- | ------------------------------ |
+| postal_code      | integer(7) | null: false                    |
+| prefecture_id    | integer    | null: false                    |
+| city             | string     | null: false                    |
+| house_number     | integer    | null: false                    |
+| phone_number     | integer    | null: false                    |
+| building_name    | string     |                                |
+| purchase_record  | references | null: false, foreign_key: true |
```
```ruby
| postal_code      | integer(7) | null: false                    |
| house_number     | integer    | null: false                    |
```
郵便番号はハイフンを必要とすることも考慮してstring型で実装しましょう！
番地に関しても、数値以外を必要とすることも考慮してstring型で実装しましょう！

```ruby
| phone_number     | integer    | null: false                    |
```
電話番号ですが、integer型に指定すると仮に「09012345678」を保存しようとした場合、先頭の0が保存されないので、string型に修正しましょう。

hi-roto replied:
修正完了しました。
f829da1

Hide-Morishita reviewed (commented):
お待たせいたしました、レビュー依頼ありがとうございます。
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いいたします。

ご質問いただきありがとうございます。
下記回答になりますので、ご確認をお願いいたします。
> アソシエーションについてですが、purchase_recordsテーブルとitemsテーブルの関係は、１対１という解釈でよろしいでしょうか？

そちらの解釈で問題ございません。

> 必須項目ではないカラムについて別テーブルを作成して、そこにデータを保存し、空白欄がないように設計したのですが、この解釈は間違っているのでしょうか？

こちらご自身で考えていただいたことは素晴らしいです。しかし、住所の情報を一つのテーブルにまとめることができる中で、あえてテーブルを分ける必要はございません。情報を管理しやすくする観点からもできるだけまとめていただくことをおすすめいたします。

> カラムを設定する際は「必須項目」が保存すべきカラムであると記載がありましたが、仮にデータベースに値を保存しない場合、実際に商品を購入して、購入者が入力した住所に配送する場合はどのように必須項目ではない（データベースに保存されない、ここで言う「建物名」）配送先を特定することになるのでしょうか？

建物名は、ビル名やアパート名などになるかと存じます。必ずしもビルやアパートの情報を入力するわけではございません。
「郵便番号」「都道府県」「市町村」「番地」を必須にすることで、配送先は特定できるかと存じます。

ご不明な点がございましたらコネクトにてご質問いただければと存じます。

hi-roto committed:
passwordカラム名の変更

nicknameカラムについても一意制限がかかっていたので、オプションに追記

hi-roto committed:
regionカラム名をprefecture_idに変更

hi-roto committed:
postal_codeとphone_numberの型をstringに変更

hi-roto commented:
レビューありがとうございます。
大変勉強になります。
最新のER図を添付しますので、よろしくお願いします。
https://gyazo.com/259f4a29d6d0984e095eb83486dbf209

tenpadrummer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+
+| Column             | Type       | Options                        |
+| -------------------| -----------| -------------------------------|
+| item_name          | string     | null: false                    |
```
こちらですが、nameだけで大丈夫です！
（item_はなくても、itemsテーブルのitemの名前ということがわかるためです）

hi-roto replied:
ご指摘ありがとうございます。
修正しました。
7d15448

tenpadrummer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+| Column             | Type       | Options                        |
+| -------------------| -----------| -------------------------------|
+| item_name          | string     | null: false                    |
+| introduct          | text       | null: false                    |
```
こまかいですが、名詞の方がいいですね。
(introduction)

hi-roto replied:
修正しました。
4ea90b8

tenpadrummer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+
+### Association
+- belongs_to :user
+- belongs_to :shipping_adress
```
- belongs_to :shipping_adress

こちら、belongs_to :itemではないでしょうか？

hi-roto replied:
その通りです。ありがとうございます。
修正しました。
0e1eef7

tenpadrummer commented on README.md:
```diff
@@ -1,24 +1,63 @@
+- belongs_to :shipping_adress
+- has_one :shipping_adress
+
+## shipping_adresses テーブル
```
_adressesのスペルを正しくしておきましょう。addressesですね。

英単語のスペルミスによるバグや致命的なミスが発生することを防ぐため、
設計段階でaddressに関わる部分を正しいスペルにしておくことをお勧めします。

hi-roto replied:
細かい点まで指導いただきありがとうございます。
設計段階で、しっかりとケアレスミスがないかどうか確認しようと思います。
修正完了しました。
5358b8c

tenpadrummer reviewed (commented):
レビュー依頼ありがとうございます！
細かい点も含めて、いくつかコメントしましたのでご確認をお願いします！

hi-roto committed:
item_nameカラムをnameカラムに変更

hi-roto committed:
introductをintroductionに変更

hi-roto committed:
purchase_recordsの関連付けをitemに変更

hi-roto committed:
adressのスペルミスを訂正

hi-roto commented:
細かい点までご指導いただき、ありがとうございます。
修正完了しました。
ER図です。
https://gyazo.com/560474c34cdede7169537710824ee778

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,63 @@
+| postal_code      | string     | null: false                    |
+| prefecture_id    | integer    | null: false                    |
+| city             | string     | null: false                    |
+| house_number     | integer    | null: false                    |
```
>| house_number     | integer    | null: false                    |

こちらintegerと記述されておりますがstring型にいたしましょう。
- `○番地〇〇`と記述する場合もあるため

hi-roto replied:
修正しました。
d512651

Katsuno0814 reviewed (commented):
レビュー依頼並びに修正ありがとうございます！
修正点確認させていただきました。
１点追加でコメントいたしましたので、ご確認お願いします！

hi-roto committed:
house_numberカラムの型名をstringに変更

hi-roto commented:
レビューありがとうございます。
修正完了しました。
ER図です。
https://gyazo.com/65f7c32cf3d4e4b74572e401e4977771

yuri240 reviewed (commented):
修正とレビュー依頼ありがとうございます。

問題なく実装できていることが確認できましたので、
LGTMとさせていただきます！

こちらのブランチは忘れずにMergeを行っていただき、
ローカルのMasterにPullしておきましょう。

データベース設計の実装、大変お疲れ様でした。
引き続き、応援しております。
